File: python/walker.py
#!/usr/bin/pypy3
from bestline import BestLine
from line import Point
from random import Random

class Walker:
	def __init__(self, n: int, k: int, random: Random):
		self.random = random
		location = Point(0, 0)
		self.marbles = []
		self.visited = set([location])
		if n == 0 or k == 0: return
		while not self.__trapped(location):
			desired = self.__validNeighbor(location)
			if len(self.visited) % n == 0:
				self.marbles.append(desired)
			self.visited.add(desired)
			location = desired
			if len(self.visited) > n * k: break
	def __trapped(self, p: Point) -> bool:
		return (
			p.up() in self.visited and
			p.right() in self.visited and
			p.down() in self.visited and
			p.left() in self.visited
		)
	def __validNeighbor(self, p: Point) -> Point:
		neighbor = self.__randomNeighbor(p)
		while neighbor in self.visited:
			neighbor = self.__randomNeighbor(p)
		return neighbor
	def __randomNeighbor(self, p: Point) -> Point:
		match self.random.randrange(4):
			case 0: return p.up()
			case 1: return p.right()
			case 2: return p.down()
			case _: return p.left()
	def intersections(self) -> int:
		return BestLine(self.marbles).numPoints

File: cpp/walker.cpp
#include "walker.hpp"    // Walker
#include "best_line.hpp" // BestLine
#include "point.hpp"     // Point
#include <ctime>         // time
#include <functional>    // bind
#include <random>        // uniform_int_distribution, default_random_engine

Walker::Walker(unsigned n, unsigned k) {
	Point location(0, 0);
	visited.insert(location);
	if (n == 0 || k == 0) return;
	while (!trapped(location)) {
		Point desired = randomNeighbor(location);
		while (visited.contains(desired)) {
			desired = randomNeighbor(location);
		}
		if (visited.size() % n == 0) {
			marbles.push_back(desired);
		}
		visited.insert(desired);
		location = desired;
		if (visited.size() > n * k) break;
	}
}

int Walker::intersections() const {
	return BestLine(marbles).intersections();
}

int Walker::steps() const {
	return visited.size();
}

Point Walker::randomNeighbor(const Point& p) const {
	static auto gen = bind(
			std::uniform_int_distribution<>(0, 3),
			std::default_random_engine(std::time(0)));
	switch (gen()) {
		case 0: return Point(p.x, p.y + 1);
		case 1: return Point(p.x + 1, p.y);
		case 2: return Point(p.x, p.y - 1);
		default: return Point(p.x - 1, p.y);
	}
}

bool Walker::trapped(const Point& p) const {
	return
		visited.contains(Point(p.x, p.y + 1)) &&
		visited.contains(Point(p.x + 1, p.y)) &&
		visited.contains(Point(p.x, p.y - 1)) &&
		visited.contains(Point(p.x - 1, p.y));
}
